### PR TITLE
Split ConstValue into two enums

### LINKED ISSUE #59210: Factor out "value" variants of `ConstValue`
Basically create a new enum `ConstKind` which has all the variants of `ConstValue` except for `Scalar`, `Slice`, `ByRef`. `ConstKind` should then have a variant (maybe `Evaluated` or just `Value`), which contains a `ConstValue`.

cc https://github.com/rust-lang/rust/pull/59178#discussion_r265554605

continuation of https://github.com/rust-lang/rust/issues/54738

This is actually just some mechanical refactoring, but also a ton of work (since you'll be touching almost every occurrence of `ConstValue`).

### PR DESCRIPTION
**Pull request by cjgillot**
Hello,

Issue #59210 appeared abandoned, so I gave it a go.
Some further cleanup and refactoring may be mandated.

I did not test beyond `x.py check`, since my home computer dies compiling librustc.

Fixes #59210

### CODE REVIEW

File: src/librustc/ty/relate.rs [hunk 1]
```diff
@@ -561,51 +561,62 @@ pub fn super_relate_consts<R: TypeRelation<'tcx>>(
...
+                }
+
+                // FIXME(const_generics): handle `ConstValue::ByRef`.
 
-        // FIXME(const_generics): handle `ConstValue::ByRef`.
+                _ =>  Err(TypeError::ConstMismatch(expected_found(relation, &a, &b))),
+            };
+
+            match new_val {
+                Ok(val) => Ok(ty::ConstKind::Value(val)),
```
Comment (Centril) [inline src/librustc/ty/relate.rs]: Could use `.map`?

File: src/librustc/ty/sty.rs [hunk 2]
```diff
@@ -2257,17 +2257,17 @@ impl<'tcx> TyS<'tcx> {
 pub struct Const<'tcx> {
     pub ty: Ty<'tcx>,
 
-    pub val: ConstValue<'tcx>,
+    pub val: ConstKind<'tcx>,
 }
 
 #[cfg(target_arch = "x86_64")]
-static_assert_size!(Const<'_>, 40);
+static_assert_size!(Const<'_>, 48);
```
Comment (Centril) [inline src/librustc/ty/sty.rs]: Aw :(

**Comment by Centril** [comment]
cc @oli-obk @varkor

**Comment by cjgillot** [comment]
I have absolutely no idea how to fix those tests.

**Comment by oli-obk** [comment]
You can run the tests with `--bless` to update the ui output to your changes. It's just some hashes changing, which is totally expected in some cases

**Comment by cjgillot** [comment]
Blessed the hashes and the mir dump.

**Pull request merged**